### PR TITLE
Add update functionality for NodeGroup Instances

### LINKED ISSUE #32: Update functionality for Nodegroup Controller
NG Controller currently supports Create and Delete operations, need to identify and implement the use cases for update.

### LINKED ISSUE #53: Allow user to specify EC2 instance type for worker nodes
At the moment nodegroup worker nodes are created of m5.large type by default https://github.com/awslabs/aws-eks-cluster-controller/blob/master/pkg/controller/nodegroup/cfn_template.go#L13 

User should be able to pass other instance type in Spec of EKS cluster to get different type of ec2 instance.

### PR DESCRIPTION
**Pull request by Magizhchi**
*Issue #, if available:*  This commit resolves #32 and resolves #53 

*Description of changes:* Added support to specify the instance details from the CRD, this PR also implements update functionality for nodegroup parameters.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go [hunk 1]
```diff
@@ -37,6 +39,24 @@ type Statement struct {
 	Resource []string `json:"resource"`
 }
 
+// Instance defines the instance details of the nodegroup
+type Instance struct {
+	// Default value is m5.large
```
Comment (anand99) [inline pkg/apis/cluster/v1alpha1/nodegroup_types.go]: I think default is t2.medium in CFN.

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go [hunk 2]
```diff
@@ -37,6 +39,24 @@ type Statement struct {
...
+	// +kubebuilder:validation:Enum=t2.small,t2.medium,t2.large,t2.xlarge,t2.2xlarge,m3.medium,m3.large,m3.xlarge,m3.2xlarge,m4.large,m4.xlarge,m4.2xlarge,m4.4xlarge,m4.10xlarge,m5.large,m5.xlarge,m5.2xlarge,m5.4xlarge,m5.12xlarge,m5.24xlarge,c4.large,c4.xlarge,c4.2xlarge,c4.4xlarge,c4.8xlarge,c5.large,c5.xlarge,c5.2xlarge,c5.4xlarge,c5.9xlarge,c5.18xlarge,i3.large,i3.xlarge,i3.2xlarge,i3.4xlarge,i3.8xlarge,i3.16xlarge,r3.xlarge,r3.2xlarge,r3.4xlarge,r3.8xlarge,r4.large,r4.xlarge,r4.2xlarge,r4.4xlarge,r4.8xlarge,r4.16xlarge,x1.16xlarge,x1.32xlarge,p2.xlarge,p2.8xlarge,p2.16xlarge,p3.2xlarge,p3.8xlarge,p3.16xlarge
+	// +optional
+	InstanceType *string `json:"instanceType,omitempty"`
+	// Default Value is 3
```
Comment (anand99) [inline pkg/apis/cluster/v1alpha1/nodegroup_types.go]: I think Default is 2 in CFN in master branch. Please update CFN to reflect ideal defaults.

File: pkg/controller/nodegroup/nodegroup_controller.go [hunk 3]
```diff
@@ -281,7 +305,100 @@ func (r *ReconcileNodeGroup) createNodeGroupStack(cfnSvc cloudformationiface.Clo
...
+			},
+		},
+		Parameters: parameters,
 	})
 
 	return err
 }
+
+func parseCFNParameterFromCRD(ngSpec clusterv1alpha1.NodeGroupSpec) []*cloudformation.Parameter {
+	if ngSpec.Instance == nil {
```
Comment (anand99) [inline pkg/controller/nodegroup/nodegroup_controller.go]: Minor: Do we have a plan to add more parameters other than instance specific? If not may be we can rephrase this as parseInstanceParameters.

File: pkg/controller/nodegroup/nodegroup_controller.go [hunk 4]
```diff
@@ -281,7 +305,100 @@ func (r *ReconcileNodeGroup) createNodeGroupStack(cfnSvc cloudformationiface.Clo
...
+		return nil
+	}
+
+	var parameter []*cloudformation.Parameter
+
+	if ngSpec.Instance.InstanceType != nil {
+		parameter = append(parameter, &cloudformation.Parameter{
+			ParameterKey:   aws.String("NodeInstanceType"),
+			ParameterValue: ngSpec.Instance.InstanceType,
```
Comment (anand99) [inline pkg/controller/nodegroup/nodegroup_controller.go]: aws.String(*parseCFNParameterFromCRD) ?

File: pkg/controller/nodegroup/nodegroup_controller.go [hunk 5]
```diff
@@ -218,7 +222,27 @@ func (r *ReconcileNodeGroup) Reconcile(request reconcile.Request) (reconcile.Res
...
+	}
+
+	if shouldUpdate(crdParameters, cfnParameters) {
+		logger.Info("Updating the NodeGroup stack with new parameters")
+		err := r.updateNodeGroupStack(cfnSvc, instance, eksCluster, crdParameters)
+		if err != nil {
+			r.fail(instance, "error updating nodegroup cloudformation stack", err, logger)
+			return reconcile.Result{}, err
+		}
+		logger.Info("cloudformation stack updated successfully")
```
Comment (anand99) [inline pkg/controller/nodegroup/nodegroup_controller.go]: Are waiting to see the StackUpdate completed in updateNodeGroupStack()?
If not then this log message is misleading.

File: pkg/controller/nodegroup/nodegroup_controller.go [hunk 6]
```diff
@@ -218,7 +222,27 @@ func (r *ReconcileNodeGroup) Reconcile(request reconcile.Request) (reconcile.Res
...
 	}
 
-	// TODO: Add parameters to obj, and update if parameters change.
+	cfnParameters, err := getCFNParametersFromCFNStack(cfnSvc, instance)
+	if err != nil {
+		r.fail(instance, "error trying to read the CFN Parameters", err, logger)
+		return reconcile.Result{}, err
+	}
+
+	if shouldUpdate(crdParameters, cfnParameters) {
```
Comment (anand99) [inline pkg/controller/nodegroup/nodegroup_controller.go]: So if User make changes to CRD for Version or IAMPolicies, the update will not be invoked. It means user has to find workaround to invoke update for  Version or IAMPolicies by updating any instance parameter. 

Can we include Version and IAMPolicies changes inclusive in shouldUpdate?

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go (hunk 2 above)
Comment (Magizhchi) [reply pkg/apis/cluster/v1alpha1/nodegroup_types.go]: Do you mean update CFN to have 3 as the default value, the Subnets are currently set to 3 so having the default value as 3 would be ideal IMO.

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go (hunk 1 above)
Comment (Magizhchi) [reply pkg/apis/cluster/v1alpha1/nodegroup_types.go]: Done

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 4 above)
Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: Can you please rephrase this comment.

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 3 above)
Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: I wanted this to reflect the work the function does, Parse the parameters for the CFN Stack  from the CRD.  The naming seems coherent with the output: []*cloudformation.Parameters.

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 5 above)
Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: Agreed, have removed the log message

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 6 above)
Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: Good call out on the Version, we have already faced a scenario where this is needed- I will submit this as a new PR as changes need to be applied for both the controlPlane and NodeGroup. 
As for the IAM policies the current way we check for updates is by reading the Parameters of the CFN Template, the IAM policies cannot be moved to the Parameters section, hence we will need to parse the CFN template body to get this value. I would defer from going down this route for now.

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go (hunk 2 above)
Comment (anand99) [reply pkg/apis/cluster/v1alpha1/nodegroup_types.go]: yes

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 6 above)
Comment (anand99) [reply pkg/controller/nodegroup/nodegroup_controller.go]: Please create an issue to address this one.

File: config/samples/cluster_v1alpha1_eks.yaml [hunk 7]
```diff
@@ -26,12 +30,3 @@ spec:
               - effect: "Allow"
                 action: ["*"]
                 resource: ["*"]
-    - name: "g2"
```
Comment (anand99) [inline config/samples/cluster_v1alpha1_eks.yaml]: May be not a bad idea to keep this in example to show it can create multiple nodegroups.

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go [hunk 8]
```diff
@@ -37,6 +39,24 @@ type Statement struct {
...
 }
 
+// Instance defines the instance details of the nodegroup
+type Instance struct {
+	// Default value is t2.medium
+	// +kubebuilder:validation:Enum=t2.small,t2.medium,t2.large,t2.xlarge,t2.2xlarge,m3.medium,m3.large,m3.xlarge,m3.2xlarge,m4.large,m4.xlarge,m4.2xlarge,m4.4xlarge,m4.10xlarge,m5.large,m5.xlarge,m5.2xlarge,m5.4xlarge,m5.12xlarge,m5.24xlarge,c4.large,c4.xlarge,c4.2xlarge,c4.4xlarge,c4.8xlarge,c5.large,c5.xlarge,c5.2xlarge,c5.4xlarge,c5.9xlarge,c5.18xlarge,i3.large,i3.xlarge,i3.2xlarge,i3.4xlarge,i3.8xlarge,i3.16xlarge,r3.xlarge,r3.2xlarge,r3.4xlarge,r3.8xlarge,r4.large,r4.xlarge,r4.2xlarge,r4.4xlarge,r4.8xlarge,r4.16xlarge,x1.16xlarge,x1.32xlarge,p2.xlarge,p2.8xlarge,p2.16xlarge,p3.2xlarge,p3.8xlarge,p3.16xlarge
+	// +optional
+	InstanceType *string `json:"instanceType,omitempty"`
+	// Default Value is 3
+	// +kubebuilder:validation:Maximum=10
```
Comment (anand99) [inline pkg/apis/cluster/v1alpha1/nodegroup_types.go]: Is there official number on how many worker nodes in EKS cluster can be supported?

File: pkg/controller/nodegroup/nodegroup_controller.go (hunk 6 above)
Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: https://github.com/awslabs/aws-eks-cluster-controller/issues/85

Comment (Magizhchi) [reply pkg/controller/nodegroup/nodegroup_controller.go]: #86

File: config/samples/cluster_v1alpha1_eks.yaml (hunk 7 above)
Comment (Magizhchi) [reply config/samples/cluster_v1alpha1_eks.yaml]: Of course, have included the same. Thanks for catching this

File: pkg/apis/cluster/v1alpha1/nodegroup_types.go (hunk 8 above)
Comment (Magizhchi) [reply pkg/apis/cluster/v1alpha1/nodegroup_types.go]: The maximum specified here is the max value for ASG as given by AWS. I couldn't find the official number of worker nodes for EKS, so did the sane thing and tested out by creating a deployment with 10 nodes, and it works 😛
